pkg/auth/requests/impersonate.go

```diff
@@ -53,48 +58,81 @@ func (h *impersonatingAuth) Authenticate(req *http.Request) (k8sUser.Info, bool,
 			if err != nil {
 				return nil, false, err
 			} else if !canDo {
-				return nil, false, errors.New("not allowed to impersonate")
+				return nil, false, errors.New("not allowed to impersonate user")
 			}
 			impersonateUser = true
-		}
 
-		if len(reqGroup) > 0 && !groupsEqual(reqGroup, groups) {
-			canDo, err := h.sar.UserCanImpersonateGroups(req, user, reqGroup)
-			if err != nil {
-				return nil, false, err
-			} else if !canDo {
-				return nil, false, errors.New("not allowed to impersonate")
+			for _, g := range reqGroup {
+				if slices.Contains(groups, g) {
+					//user belongs to the group they are trying to impersonate
+					continue
+				}
+				canDo, err := h.sar.UserCanImpersonateGroup(req, user, g)
+				if err != nil {
+					return nil, false, err
```

Comment:
Should we wrap the error? E.g. something along the lines of
```suggestion
					return nil, false, fmt.Errorf("error checking if user can impersonate group: %w", err)
```